app/src/main/cpp/skyline/vfs/backing.h

```diff
@@ -42,13 +53,37 @@ namespace skyline::vfs {
          * @param offset The offset to start reading from
          * @return The amount of bytes read
          */
-        virtual size_t Read(span <u8> output, size_t offset = 0) = 0;
+        size_t ReadUnchecked(span <u8> output, size_t offset = 0) {
+            if (!mode.read)
+                throw exception("Attempting to read a backing that is not readable");
+
+            return ReadImpl(output, offset);
+        };
+
+        /**
+         * @brief Read bytes from the backing at a particular offset to a buffer and check to ensure the full size was read
+         * @param output The object to write the data read to
+         * @param offset The offset to start reading from
+         * @return The amount of bytes read
+         */
+        size_t Read(span <u8> output, size_t offset = 0) {
+            if (offset > size)
+                throw exception("Offset cannot be past the end of a backing");
+
+            if ((size - offset) < output.size())
+                throw exception("Trying to read past the end of a backing: 0x{:X}/0x{:X} (Offset: 0x{:X})", output.size(), size, offset);
+
+            if (ReadUnchecked(output, offset) != size)
+                throw exception("Failed to read the requested size from backing");
+
+            return size;
+        };
```

Comment:
I see that you've done the most optimal thing with the API here :stuck_out_tongue:

Reply:
🙏